typhon/engine.py

```diff
@@ -0,0 +1,226 @@
+        self.addItems([self.null_command] + self.available_commands[state])
+
+
+class EngineWidget(QGroupBox):
+    """
+    RunEngine Control Widget
+
+    Parameters
+    ----------
+    engine : RunEngine, optional
+        The underlying RunEngine object. A basic version wil be instatiated if
+        one is not provided
+
+    plan : callable, optional
+        A callable  that takes no parameters and returns a generator. If the
+        plan is meant to be called repeatedly the function should make sure
+        that a refreshed generator is returned each time
+
+    Attributes
+    ----------
+    engine_state_change : pyqtSignal('QString', 'QString')
+        Signal emitted by changes in the RunEngine state. The first string is
+        the current state, the second is the previous state
+
+    update_rate: float
+        Update rate the qt_kicker is installed at
+
+    command_registry: dict
+        Mapping of commands received by the pyqtSlot `command` and actual
+        Python callables
+    """
+    engine_state_change = pyqtSignal('QString', 'QString')
+    update_rate = 0.02
+    command_registry = {'Halt': RunEngine.halt,
+                        'Start': no_plan_warning,
+                        'Abort': RunEngine.abort,
+                        'Resume': RunEngine.resume,
+                        'Pause': RunEngine.request_pause,
+                        'Stop': RunEngine.stop}
+
+    def __init__(self, engine=None, plan=None, parent=None):
+        # Instantiate widget information and layout
+        super().__init__('Engine Control', parent=parent)
+        self.setStyleSheet('QLabel {qproperty-alignment: AlignCenter}')
+        self.label = EngineLabel(parent=self)
+        self.command_label = QLabel('Available Commands')
+        self.status_label = QLabel('Engine Status')
+        self.control = EngineControl()
+        lay = QVBoxLayout()
+        lay.addWidget(self.status_label)
+        lay.addWidget(self.label)
+        lay.addWidget(self.command_label)
+        lay.addWidget(self.control)
+        self.setLayout(lay)
+        # Create a new RunEngine if we were not provided one
+        self._engine = None
+        self._plan = None
+        self.engine = engine or RunEngine()
+
+    @property
+    def plan(self):
+        """
+        Stored plan callable
+        """
+        return self._plan
+
+    @plan.setter
+    def plan(self, plan):
+        logger.debug("Storing a new plan for the RunEngine")
+        # Do not allow plans to be set while RunEngine is active
+        if self.engine and self.engine.state != 'idle':
+            logger.exception("Can not change the configured plan while the "
+                             "RunEngine is running!")
+            return
+        # Store our plan internally
+        self._plan = plan
+        # Register a new call command
+        self.command_registry['Start'] = (lambda x:
+                                          RunEngine.__call__(x, self.plan()))
```

Comment:
If I understand this correctly, you have to touch the right properties in the right order (engine -> plan), which modifies the command_registry dict, making up a new function to call the plan when 'Start' is selected.

This seems a bit over-complicated...  How about instead registering 'Start' as an instance method on the widget? Failing that, special-casing in 'command' below?

Reply:
Interesting. I'll think about this. I also just realized that this setup enforces an order of operations in setting the plan and the engine. This absolutely needs to be corrected.